spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,362 @@
+  // }
+
+  // Once command streaming is implemented, this initialization should go here.
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+std::vector<hardware_interface::StateInterface> SpotHardware::export_state_interfaces() {
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (size_t i = 0; i < info_.joints.size(); i++) {
+    const auto& joint = info_.joints.at(i);
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                     &hw_states_[interfaces_per_joint_ * i]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 1]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 2]));
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface> SpotHardware::export_command_interfaces() {
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto& joint = info_.joints.at(i);
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                         &hw_commands_[interfaces_per_joint_ * i]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 1]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 2]));
+  }
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // Once command streaming is enabled, this should release the lease and stop command streaming
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_shutdown(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  // Once command streaming is enabled, this should also release the lease and stop command streaming
+  // (if not already deactivated)
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::return_type SpotHardware::read(const rclcpp::Time& /*time*/, const rclcpp::Duration& /*period*/) {
+  state_streaming_handler_.get_joint_states(joint_states_);
```

Comment:
In order to do this, we would need to know the correct number of joints so that the vectors are the right size, right? 
Right now there is not a distinction between robots with/without arms (19 vs 12 joints). We could add a check for this (like has_arm) but I don't want to make this PR even bigger

Reply:
Isn't that information implied in `hw_states_` size?

Reply:
Oops, you are right, updated